pyo3-derive-backend/src/frompy.rs

```diff
@@ -0,0 +1,428 @@
+                }
+            }
+            Fields::Unit => {
+                return Err(syn::Error::new_spanned(
+                    &fields,
+                    "Cannot derive FromPyObject for Unit structs and variants",
+                ))
+            }
+        };
+        let err_name = maybe_renamed_err(&attrs)?
+            .map(|s| s.value())
+            .unwrap_or_else(|| ident.to_string());
+
+        let v = Container {
+            ident: &ident,
+            style,
+            err_name,
+        };
+        Ok(v)
+    }
+
+    /// Build derivation body for a struct.
+    fn derive_struct(&self) -> TokenStream {
+        match &self.style {
+            Style::StructNewtype(ident) => self.build_newtype_struct(Some(&ident)),
+            Style::TupleNewtype => self.build_newtype_struct(None),
+            Style::Tuple(len) => self.build_tuple_struct(*len),
+            Style::Struct(tups) => self.build_struct(tups),
+        }
+    }
+
+    fn build_newtype_struct(&self, field_ident: Option<&Ident>) -> TokenStream {
+        if let Some(ident) = field_ident {
+            quote!(
+                Ok(Self{#ident: obj.extract()?})
+            )
+        } else {
+            quote!(Ok(Self(obj.extract()?)))
+        }
+    }
+
+    fn build_tuple_struct(&self, len: usize) -> TokenStream {
+        let mut fields: Punctuated<TokenStream, syn::Token![,]> = Punctuated::new();
+        for i in 0..len {
+            fields.push(quote!(slice[#i].extract()?));
+        }
+        quote!(
+            let s = <::pyo3::types::PyTuple as ::pyo3::conversion::PyTryFrom>::try_from(obj)?;
+            let seq_len = s.len();
+            if seq_len != #len {
+                let msg = format!(
+                    "Expected tuple of length {}, but got length {}.",
+                    #len,
+                    seq_len
+                );
+                return Err(::pyo3::exceptions::PyValueError::py_err(msg))
+            }
+            let slice = s.as_slice();
+            Ok(Self(#fields))
+        )
+    }
+
+    fn build_struct(&self, tups: &[(&Ident, syn::ExprCall)]) -> TokenStream {
+        let mut fields: Punctuated<TokenStream, syn::Token![,]> = Punctuated::new();
+        for (ident, ext_fn) in tups {
+            fields.push(quote!(#ident: obj.#ext_fn?.extract()?));
+        }
+        quote!(Ok(Self{#fields}))
+    }
+
+    fn check_transparent_len(fields: &Fields) -> Result<()> {
+        if fields.len() != 1 {
+            return Err(syn::Error::new_spanned(
+                fields,
+                "Transparent structs and variants can only have 1 field",
+            ));
+        }
+        Ok(())
+    }
+}
+
+/// Get the extraction function that's called on the input object.
+///
+/// Valid arguments are `get_item`, `get_attr` which are called with the
+/// stringified field identifier or a function call on `PyAny`, e.g. `get_attr("attr")`
+fn ext_fn(attrs: &[Attribute], field_ident: &Ident) -> Result<syn::ExprCall> {
```

Comment:
For attributes, it may be nice to have an `enum` which encodes all the attributes, along with a `parse_derive_attributes` function which returns a `Vec` of the attributes found.

Awkward cases that might need to be handled: 
- what happens if the same attribute is parsed twice?
- what happens if the attribute is provided in an invalid context? E.g. `extract` attribute in a tuple struct - would probably be desirable to raise a helpful error.